client/services/code_preprocessors/PythonCodePreprocessorService.js

```diff
@@ -152,117 +160,183 @@ tie.factory('PythonCodePreprocessorService', [
+            VARNAME_BUGGY_OUTPUT_TEST_RESULTS +
+            '.append('+ VARNAME_TASK_BUGGY_OUTPUT_TEST_RESULTS +')\n');
+        });
+      }
 
       return fullTestCode;
     };
 
-    var _generatePerformanceTestCode = function(performanceTests) {
-      var testCode = '';
-
-      performanceTests.forEach(function(test, index) {
-        var qualifiedEvaluationFunctionName = (
-          CLASS_NAME_STUDENT_CODE + '().' + test.getEvaluationFunctionName());
-        var qualifiedTransformationFunctionName = (
-          test.getTransformationFunctionName());
-        // TODO(eyurko): Make this work for non-linear runtimes, such as log(n).
-        // TODO(eyurko): Use linear regression to determine if the data points
-        // "look" linear, quadratic, etc, and then provide feedback accordingly.
-        testCode = [
-          '',
-          'def get_test_input(atom, input_size):',
-          '    return ' + qualifiedTransformationFunctionName + (
-            '(atom, input_size)'),
-          '',
-          'def run_performance_test(test_input):',
-          '    time_array = []',
-          '    for input_size in [' + SMALL_INPUT_SIZE + (
-            ', ' + LARGE_INPUT_SIZE + ']:'),
-          '        start = time.clock()',
-          '        output = ' + qualifiedEvaluationFunctionName + (
-            '(get_test_input(test_input, input_size))'),
-          '        finish = time.clock() - start',
-          '        time_array.append(finish)',
-          '    if time_array[1] > ' + UPPER_BOUND_RATIO_IF_LINEAR + (
-            ' * time_array[0]:'),
-          '        return "not linear"',
-          '    return "linear"',
-          '',
-          VARNAME_PERFORMANCE_TEST_RESULTS + ' = []'
-        ].join('\n');
+    var _generatePerformanceTestCode = function(allTasksPerformanceTests) {
+      var testCode = (VARNAME_PERFORMANCE_TEST_RESULTS + ' = []\n');
+      var performanceTestFunctionSet = new Set();
+      allTasksPerformanceTests.forEach(function(oneTaskPerformanceTests, index) {
+        testCode += (VARNAME_TASK_PERFORMANCE_TEST_RESULTS + ' = []\n')
+        oneTaskPerformanceTests.forEach(function(test) {
+          var qualifiedEvaluationFunctionName = (
+            CLASS_NAME_STUDENT_CODE + '().' + test.getEvaluationFunctionName());
+          var qualifiedTransformationFunctionName = (
+            test.getTransformationFunctionName());
+          // TODO(eyurko): Make this work for non-linear runtimes, such as log(n).
+          // TODO(eyurko): Use linear regression to determine if the data points
+          // "look" linear, quadratic, etc, and then provide feedback accordingly.
+          if (!performanceTestFunctionSet.has(qualifiedEvaluationFunctionName)) {
+            performanceTestFunctionSet.add(qualifiedEvaluationFunctionName);
+            testCode += [
+              '',
+              'def get_test_input_from_'+
+                  qualifiedTransformationFunctionName.replace('.', '_') +
```

Comment:
Same thing -- this sort of munging of the function names is a bit hacky. Can we treat the functions as first-class objects and pass them in as args instead?